### PR TITLE
Elections

### PR DESCRIPTION
**Pull request by MartinBriand**
Added all the election interface
Better sidebar

Two things are lacking : 
- selecting users when creating elections
- perhaps loading all users for every election while listing might get long, it could be great to create another serializer without the list and then when clicking the "voir les personnes autorisées à participer à l'élection", load the list

### CODE REVIEW

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    const maxChoicesPerBallot = parseInt(election.maxChoicesPerBallot);
+
+    const [selected, setSelected] = useState<number[]>([]);
+    const [showVoteConfirmation, setShowVoteConfirmation] = useState<boolean>(false);
+    const newToast = useContext(ToastContext);
+
+    const disabledVoteButton = disabled || selected.length===0;
+
+
+    const handleClick = (choice: Choice) => {
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/ActiveList.tsx:98]: Please add one or two comments to quickly explain what is going on here

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                <Row>
+                    <ListOfVotersButtonModal election={election}/>
+                </Row>
+            </Card.Body>
+        </Card>
+    )
+};
+
+const Vote = ({ election, association, ...props }) => {
+    const isRegisterd = election.isRegistered;
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/ActiveList.tsx:86]: Typo here I guess?

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            </Button>
+        )
+    } else if (!props.isRegistered) {
+        return (
+            <Button
+                className={"btn btn-success float-right mt-5"}
+                block
+                disabled
+            >
+                Vous n'êtes inscrit à ce scrutin
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/ActiveList.tsx:197]: pas*

File: frontend/src/services/apiService.ts [hunk 4]
```diff
@@ -5,9 +5,17 @@ import { Page } from "../models/associations/page";
 import { Marketplace, Transaction } from "../models/associations/marketplace";
 import { Media } from "../models/associations/media";
 import { QueryResult, useQuery } from "react-query";
+import {Ballot, Election, Result} from "../models/associations/election";
+import {User} from "../models/user";
 
 const baseApi = "http://localhost:8000/api/v1";
 
+export enum activeStatus {
```
Comment (Piwakk) [inline frontend/src/services/apiService.ts:13]: Please give it a more precise name, like `ElectionStatus`.
Plus, according to the Typescript convention, the name of the enumeration should be capitalized.

File: frontend/src/pages/associations/elections/Edit.tsx [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+        <Button
+            variant={'outline-dark'}
+            onClick={() => props.onClick()}
+        >
+            Nouvelle option
+        </Button>
+    )
+};
+
+const toGoodString = (date: Date) => {
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/Edit.tsx:367]: You should take a look at [String#padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart) ;)

File: frontend/src/pages/associations/elections/Edit.tsx [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+            <>
+                <PageTitle className={"mt-6"}>Modification d'une élection</PageTitle>
+                <p>Attention ! C'est la supprimer et la remplacer. Tous les votes seront perdus. </p>
+                <EditElection {...props} election={election} association={association} edit={true}/>
+            </>
+        );
+    return null;
+};
+
+const EditElection = ({ election, association, ...props }) => {
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/Edit.tsx:56]: I am sorry, I think I haven't told you: we use [Formik](https://jaredpalmer.com/formik/docs/overview) for our forms. Another library, but it is really useful, especially for handling errors.

Some work is needed to make it work cleanly with tabler.io, and I have created some components in my branch `pollsFrontend`. We can talk about how to (re)use them.

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                </Row>
+            </>
+        )
+    }
+};
+
+const ElectionCard = ({ election, association }) => {
+    let editButton;
+    if (association.myRole.electionPermission) {
+        editButton = (
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/ActiveList.tsx:46]: Why not using a `<Button>` ?
Also, I think it would look better without the text

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+    };
+
+    const handleCloseAlert = () => {
+        setShowVoteConfirmation(false)
+    };
+
+    return (
+        <>
+            {election.choices.map(choice => (
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/ActiveList.tsx:140]: This should be embedded in a `<Form>` and used with `Formik`.

Comment (Piwakk) [reply frontend/src/pages/associations/elections/ActiveList.tsx:140]: (see on a comment below what Formik is)

File: frontend/src/pages/associations/elections/Commons.tsx [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import {Link} from "react-router-dom";
+import {Button, Modal} from "react-bootstrap";
+import Card from "react-bootstrap/Card";
+import Row from "react-bootstrap/Row";
+import {activeStatus, api, useBetterQuery} from "../../../services/apiService";
+import {LoadingAssociation} from "../Loading";
+import {Election, Ballot, Choice} from "../../../models/associations/election";
+import {ToastContext} from "../../../utils/Toast";
+
+export const ListOfVotersButtonModal = ({election, ...props}) => {
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/Commons.tsx:12]: If `Commons` only contains one Element, you could also rename the file after it

File: frontend/src/pages/associations/elections/ResultsList.tsx [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        ));
+        const startsAt = new Date(election.startsAt);
+        const endsAt = new Date(election.endsAt);
+        return (
+            <Card>
+                <Card.Body>
+                    <Row>
+                        <h4>{election.name}</h4>
+                        <p>Cette élection s'est tenu entre le {startsAt.toLocaleDateString()} à {startsAt.toLocaleTimeString()} et le {endsAt.toLocaleDateString()} à {endsAt.toLocaleTimeString()}</p>
+                        <p>Il était de possible de faire {election.maxChoicesPerBallot} choix parmi ceux proposés</p>
```
Comment (Piwakk) [inline frontend/src/pages/associations/elections/ResultsList.tsx:75]: typo

File: backend/associations/serializers/election.py [hunk 11]
```diff
@@ -151,6 +161,11 @@ def update(self, election, validated_data):
 
     def create(self, validated_data):
         registered_voters = validated_data.pop("registered_voters")
+        print("!!!!!!!!!!!!!!!!!!!!")
+        print("coucou, on est dans create")
```
[the remark this answers is not in the document]
Comment (Piwakk) [reply backend/associations/serializers/election.py:165]: Just in case that was not explicit enough: please remove it. ;)

**Comment by MartinBriand** [comment]
Hi, 
I tried to change all of what you asked but a few things are difficult to correct :

First of all, I removed all the prints and console.log, the code looks much cleaner now. Similarly, I changed all the variable's names that were not specific enough and URLs are in French now.

I had the feeling you did not have a look at the results page as well as the upcoming page. I added fixtures so that you can see these results too. You juste need to repopulate the database.

As for formik. I didn't know how to use it so I created my own form. The example of 'pages' is actually very simple and I wanted to add some verifications or complex selections (variable number of choices, selecting from list of users to register voters (not done yet but later)... If you think formik may do a better job or is more consistent with the rest of the project, I would be happy to learn about how to create custom form component with this tool. For the moment I kept with my old solution.

The problem of time picker is hard to solve. I tested on edge and chrome and it worked well. But with Firefox (I didn't try before), I don't know why, dates are not pretty and the tools to select a date doesn't work as well.

As for the design, I tried to make things look prettier, I hope you'll think the same.

Thanks
Martin

File: backend/associations/models/election.py [hunk 12]
```diff
@@ -39,14 +39,27 @@ class Election(models.Model):
...
         # We then wrap all of these choices into a Counter, which is a nice built-in object which will
         # count the ballots for us.
-        return Counter(
-            [ballot[0] for ballot in self.ballots.values_list("choices__name")]
-        )
+        return {
+            "result": Counter(
+                [ballot[0] for ballot in self.ballots.values_list("choices__id")]
+            ),
+            "nb_voters": len(self.voters.all()),
```
[the remark this answers is not in the document]
Comment (MartinBriand) [reply backend/associations/models/election.py]: done

File: backend/associations/models/election.py [hunk 13]
```diff
@@ -39,14 +39,27 @@ class Election(models.Model):
         default=1, help_text="The number of choices allowed per ballot."
     )
 
+    @cached_property
+    def active_status(self):
```
[the remark this answers is not in the document]
Comment (MartinBriand) [reply backend/associations/models/election.py:43]: I do not understand what you mean... The other cached_properties are written in lower case. isn't it ?

File: backend/associations/serializers/election.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MartinBriand) [reply backend/associations/serializers/election.py]: done

File: backend/associations/views/election.py [hunk 14]
```diff
@@ -20,18 +22,34 @@
         * Destroy an election:          DELETE  /associations/elections/1/
         * Vote to an election:          POST    /associations/elections/1/vote/
         * Results of an election:       GET     /associations/elections/1/results/
+        * List living elections:        GET     /associations/elections/living
```
[the remark this answers is not in the document]
Comment (MartinBriand) [reply backend/associations/views/election.py]: this were old urls, now, it goes through active_status url get component

File: backend/associations/views/election.py [hunk 15]
```diff
@@ -20,18 +22,34 @@
...
     serializer_class = ElectionSerializer
     permission_classes = (ElectionPermission,)
+    filterset_fields = ("association",)
+    ordering = ["-ends_at"]
+
+    def filter_queryset(self, queryset):
+        """Sorting by property (here only active_status)"""
+        active_status = self.request.query_params.get("active_status", None)
+        if active_status:
+            L = []
```
[the remark this answers is not in the document]
Comment (MartinBriand) [reply backend/associations/views/election.py]: done

File: backend/associations/views/election.py [hunk 16]
```diff
@@ -53,12 +71,16 @@ def create(self, request, *args, **kwargs):
 
         # Serialize the data.
         serializer = self.get_serializer(data=request.data)
+        print("salut")
```
[the remark this answers is not in the document]
Comment (MartinBriand) [reply backend/associations/views/election.py]: done

File: frontend/package.json [hunk 17]
```diff
@@ -14,7 +14,9 @@
         "react-router-dom": "latest",
         "react-scripts": "^3.4.1",
         "tabler-ui": "0.0.34",
-        "yup": "^0.27.0"
+        "yup": "^0.27.0",
+        "react-chartkick": "latest",
+        "chart.js": "latest"
```
[the remark this answers is not in the document]
Comment (MartinBriand) [reply frontend/package.json]: done

File: frontend/src/pages/associations/elections/ActiveList.tsx (hunk 1 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/ActiveList.tsx:100]: done

File: frontend/src/pages/associations/elections/ActiveList.tsx (hunk 2 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/ActiveList.tsx]: done

File: frontend/src/pages/associations/elections/ActiveList.tsx (hunk 3 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/ActiveList.tsx]: done

File: frontend/src/pages/associations/elections/Edit.tsx (hunk 5 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/Edit.tsx:371]: It helped so much, thanks

File: frontend/src/pages/associations/elections/ActiveList.tsx (hunk 8 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/ActiveList.tsx:142]: I didn't ;(

File: frontend/src/pages/associations/elections/Commons.tsx (hunk 9 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/Commons.tsx:18]: It contains more than one element now ;)

File: frontend/src/pages/associations/elections/ResultsList.tsx (hunk 10 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/ResultsList.tsx]: I changed the ui so it no longer exists

File: backend/associations/models/election.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Piwakk) [reply backend/associations/models/election.py:43]: I was talking about `upcoming`, `current`…

File: frontend/src/pages/associations/elections/ActiveList.tsx (hunk 7 above)
Comment (Piwakk) [reply frontend/src/pages/associations/elections/ActiveList.tsx:46]: OK, that'll do for now

**Comment by Piwakk** [comment]
Thanks for the quick corrections. I'll have a look at your changes.

File: frontend/public/index.html [hunk 18]
```diff
@@ -24,20 +24,26 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
+    <link
+            rel="stylesheet"
+            href="https://maxcdn.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
+            integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T"
+            crossorigin="anonymous"
+    />
```
Comment (abocquet) [inline frontend/public/index.html:32]: Please do not import a new CSS library, tabler.io should be able to meet all of our needs. What component did you want  to import specifically?

File: frontend/src/models/associations/election.ts [hunk 19]
```diff
@@ -2,23 +2,33 @@ import { User } from "../user";
...
+    numberOfVoters:string;
+    numberOfRegistered:string;
 }
 
 export interface Ballot {
-    id: string;
+    id?: string;
     election: Election;
-    choices: Choice[];
+    choices: Choice[] | number[];
```
Comment (abocquet) [inline frontend/src/models/associations/election.ts:20]: Why do you use type union here? We should be confident with the data types we have in the models

File: frontend/src/pages/associations/elections/Commons.tsx [hunk 20]
```diff
@@ -0,0 +1,53 @@
+import React, {useState} from "react";
+import {Button, Modal} from "react-bootstrap";
+
+export const toFrenchTimeNoSecondString = (date : Date) => {
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/Commons.tsx:4]: Take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString

File: frontend/src/pages/associations/elections/Commons.tsx [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    return date.getHours().toString().padStart(2, '0') + ":" + date.getMinutes().toString().padStart(2, '0')
+};
+
+export const toFrenchDate = (date : Date) => {
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/Commons.tsx:8]: Take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString

File: frontend/src/pages/associations/elections/Edit.tsx [hunk 22]
```diff
@@ -0,0 +1,398 @@
...
+            errors.name = 'Le titre ne peut pas être vide';
+        }
+        if (values.dates.startsAt >= values.dates.endsAt) {
+            errors.dates = "L'élection ne peut pas se terminer avant de commencer"
+        }
+        if (values.choices.includes("")) {
+            errors.choices = 'Un choix ne peut pas être vide'
+        }
+        if (values.maxChoicesPerBallot < 1) {
+            errors.maxChoicesPerBallot = 'Minimum 1'
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/Edit.tsx:132]: Could be more explicit

Comment (abocquet) [inline frontend/src/pages/associations/elections/Edit.tsx:132]:
```suggestion
            errors.maxChoicesPerBallot = 'Les électeurs doivent pouvoir faire au moins un choix par ballot'
```

File: frontend/src/index.tsx [hunk 23]
```diff
@@ -1,10 +1,10 @@
...
-import App from './App';
-import * as serviceWorker from './serviceWorker';
+import React from "react";
+import ReactDOM from "react-dom";
+import "./index.css";
+import App from "./App";
+import * as serviceWorker from "./serviceWorker";
 
-ReactDOM.render(<App />, document.getElementById('root'));
+ReactDOM.render(<App />, document.getElementById("root"));
```
Comment (abocquet) [inline frontend/src/index.tsx:7]: Please roll back these changes, they don't belong here. Maybe change the configuration of your linter

File: frontend/src/models/associations/election.ts [hunk 24]
```diff
@@ -2,23 +2,33 @@ import { User } from "../user";
...
+    name: string;
+    election: Election | number;
+}
+
+export interface Result {
     id: string;
-    election: Election;
+    result: string;
+    numberOfVoters:string;
+    numberOfRegistered:string;
```
Comment (abocquet) [inline frontend/src/models/associations/election.ts:14]: Run prettier here

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+        editButton = (
+            <Link
+                to={`/associations/${association.id}/elections/${election.id}/modifier`}
+                className={"float-right mt-0"}
+            >
+                <i className={"fe fe-edit"}/>
+            </Link>
+        );
+    } else {
+        editButton = (<></>)
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/ActiveList.tsx:54]:
```suggestion
        editButton = </>;
```

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+    }
+    const startsAt = new Date(election.startsAt);
+    const endsAt = new Date(election.endsAt);
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/ActiveList.tsx:57]: Why using an intermediary variable here?

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+                        <h4>{election.name}</h4>
+                    </Col>
+                    <Col>
+                        {editButton}
+                    </Col>
+                </Row>
+                <Row>
+                    <p>
+                        Ouverture le {toFrenchDate(startsAt)} à {toFrenchTimeNoSecondString(startsAt)}<br/>
+                        Fermeture le {toFrenchDate(endsAt)} à {toFrenchTimeNoSecondString(endsAt)}
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/ActiveList.tsx:72]:
```suggestion
                        Fermeture le {toFrenchDate(election.endsAt)} à {toFrenchTimeNoSecondString(election.endsAt)}
```

File: frontend/src/pages/associations/elections/ActiveList.tsx [hunk 28]
```diff
@@ -0,0 +1,241 @@
...
+                    <Col xs={0}>
+                        <h4>{election.name}</h4>
+                    </Col>
+                    <Col>
+                        {editButton}
+                    </Col>
+                </Row>
+                <Row>
+                    <p>
+                        Ouverture le {toFrenchDate(startsAt)} à {toFrenchTimeNoSecondString(startsAt)}<br/>
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/ActiveList.tsx:71]:
```suggestion
                        Ouverture le {toFrenchDate(election.startsAt)} à {toFrenchTimeNoSecondString(election.startsAt)}<br/>
```

File: frontend/src/pages/associations/elections/Edit.tsx [hunk 29]
```diff
@@ -0,0 +1,398 @@
...
+
+
+
+    const formIsValid = values => {
+        interface Errors {
+            name:string | null,
+            dates: string | null
+            choices:string | null
+            maxChoicesPerBallot:string|null
+            registeredVoters:string|null
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/Edit.tsx:113]: Run prettier here

File: frontend/src/pages/associations/elections/Edit.tsx [hunk 30]
```diff
@@ -0,0 +1,398 @@
...
+        </Card>
+    );
+};
+
+const EditElectionForm = ({election, association, ...props}) => {
+    const newToast = useContext(ToastContext);
+
+    const listOfChoicesNames = (choices: Choice[]) => {
+        const result: string[] = [];
+        choices.map((choice) => result.push(choice.name));
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/Edit.tsx:101]:
```suggestion
        const result = choices.map((choice) => choice.name);
```

File: frontend/src/pages/associations/elections/UpcomingList.tsx [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+        editButton = (
+            <Link
+                to={`/associations/${association.id}/elections/${election.id}/modifier`}
+                className={"float-right mt-0"}
+            >
+                <i className={"fe fe-edit"}/>
+            </Link>
+        );
+    } else {
+        editButton = (<></>)
```
Comment (abocquet) [inline frontend/src/pages/associations/elections/UpcomingList.tsx:57]:
```suggestion
        editButton = </>;
```

File: frontend/src/serviceWorker.ts [hunk 32]
```diff
@@ -11,133 +11,134 @@
...
 
 const isLocalhost = Boolean(
-  window.location.hostname === 'localhost' ||
-    // [::1] is the IPv6 localhost address.
-    window.location.hostname === '[::1]' ||
-    // 127.0.0.1/8 is considered localhost for IPv4.
-    window.location.hostname.match(
-      /^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
-    )
+    window.location.hostname === "localhost" ||
```
Comment (abocquet) [inline frontend/src/serviceWorker.ts:14]: Theses changes don't belong in this PR, please roll back

File: frontend/src/utils/random.ts [hunk 33]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Return a random item from `array`.
+ */
+export function getRandom(array: Array<any>): any {
```
Comment (abocquet) [inline frontend/src/utils/random.ts:4]: A use of templates is more relevant here
```suggestion
export function getRandom<T>(array: Array<T>): T {
```

File: frontend_legacy/package.json [hunk 34]
```diff
@@ -23,42 +23,42 @@
     "@ng-bootstrap/ng-bootstrap": "^4.2.2",
     "@ng-select/ng-select": "^2.20.5",
     "angular-markdown-editor": "^2.0.2",
-    "angular-notifier": "^4.1.1",
+    "angular-notifier": "^4.1.2",
```
Comment (abocquet) [inline frontend_legacy/package.json:26]: Don't change `frontend_legacy`, it's here only for information

File: frontend/src/models/associations/election.ts (hunk 19 above)
Comment (MartinBriand) [reply frontend/src/models/associations/election.ts:20]: Hi, I had a problem because what I receive from python and what I send are different things...
For example, here in `Ballot`, I receive `Choice`s, but when someone votes, he sends only the ids of the choice, hence the `number[]`.
Similarly, in `Election`, I receive `Choice`s, but when I create an election, I have to send an `Object` with only a `name` field, hence the `Object[]`.
Again for `registeredVoters` in `Election`, I send only their ids as a `string[]`

(I removed one which was useless

File: frontend/src/models/associations/election.ts (hunk 24 above)
Comment (MartinBriand) [reply frontend/src/models/associations/election.ts:14]: What do you mean ?

File: frontend/src/pages/associations/elections/ActiveList.tsx (hunk 26 above)
Comment (MartinBriand) [reply frontend/src/pages/associations/elections/ActiveList.tsx:58]: Because election.startsAt is a `string` I have to change to a `Date`
